proxy/ptunnel/commands/tunnel.go

```diff
@@ -138,24 +182,69 @@ func (r *Tunnel) Run(cmd *cobra.Command, args []string) (err error) {
 	return err
 }
 
-func (r *Tunnel) RunListener(listener net.Listener, proxy *url.URL, host string, port uint16, cookie *http.Cookie, hostport string) error {
-	defer listener.Close()
-	for {
-		generic, err := listener.Accept()
-		if err != nil {
-			return err
+// listenerChan returns a channel that contains connections that the Listener is
+// accepting. If the listener encounters an error, it will fill in `e` before
+// closing the returned channel and terminating.
+func listenerChan(l net.Listener, e *error) <-chan net.Conn {
+	c := make(chan net.Conn)
+	go func() {
+		defer close(c)
+		for {
+			conn, err := l.Accept()
+			if err != nil {
+				*e = err
+				return
+			} else {
+				c <- conn
+			}
 		}
+	}()
+	return c
+}
 
-		conn := generic.(*net.TCPConn)
-		id := fmt.Sprintf("tunnel by %s on %s with %s:%d via %s from %s", r.Username(), hostport, host, port, proxy, conn.RemoteAddr())
-		r.Log.Infof("%s - accepted connection", id)
-		go func() {
-			defer conn.Close()
-			err := r.RunTunnel(proxy, id, host, port, cookie, knetwork.ReadOnlyClose(conn), knetwork.WriteOnlyClose(conn))
-			if err != nil {
-				r.Log.Infof("%s - terminated with %v", id, err)
+func (r *Tunnel) RunListener(ctx context.Context, listener net.Listener, proxy *url.URL, host string, port uint16, cookie *http.Cookie, localAddr string) error {
+	var listenerErr error
+	conns := listenerChan(listener, &listenerErr)
+	for {
+		select {
+		case <-ctx.Done():
+			listener.Close()
+			// Listener is closed; drain the channel until the listener loop
+			// closes it.
+			for len(conns) > 0 {
+				<-conns
 			}
-		}()
+			// The listener will report an error from Accept because we closed
+			// it, but this is expected and can be ignored.
+			return nil
+
+		case conn, ok := <-conns:
+			// Listener loop closed the channel unexpectedly, so no more
+			// connections, and there is probably an error.
+			if !ok {
+				return listenerErr
```

Comment:
on Done, listener.Close() is called. Should the same happen here? Eg, if Accept() returns error, channel is closed, we end up here, but nobody cleans/closes the listener?